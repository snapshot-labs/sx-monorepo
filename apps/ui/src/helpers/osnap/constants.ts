// contract addresses pulled from https://github.com/UMAprotocol/protocol/tree/master/packages/core/networks
export const contractData = [
  {
    // mainnet
    network: '1',
    name: 'OptimisticOracleV3',
    address: '0xfb55F43fB9F48F63f9269DB7Dde3BbBe1ebDC0dE',
    subgraph:
      'https://subgrapher.snapshot.org/subgraph/arbitrum/Bm3ytsa1YvcyFJahdfQQgscFQVCcMvoXujzkd3Cz6aof',
    deployBlock: 16636058
  },
  {
    // goerli
    network: '5',
    name: 'OptimisticOracleV3',
    address: '0x9923D42eF695B5dd9911D05Ac944d4cAca3c4EAB',
    subgraph:
      'https://api.thegraph.com/subgraphs/name/md0x/goerli-optimistic-oracle-v3',
    deployBlock: 8497481
  },
  {
    // optimism
    network: '10',
    name: 'OptimisticOracleV3',
    address: '0x072819Bb43B50E7A251c64411e7aA362ce82803B',
    subgraph:
      'https://subgrapher.snapshot.org/subgraph/arbitrum/FyJQyV5TLNeowZrL6kLUTB9JNPyWQNCNXJoxJWGEtBcn',
    deployBlock: 74537234
  },
  {
    // gnosis
    network: '100',
    name: 'OptimisticOracleV3',
    address: '0x22A9AaAC9c3184f68C7B7C95b1300C4B1D2fB95C',
    subgraph:
      'https://subgrapher.snapshot.org/subgraph/arbitrum/9K2nctaB2rAh7Cgzx3wKtdHwWoEeEQ9AThGATak6Ngm9',
    deployBlock: 27087150
  },
  {
    // polygon
    network: '137',
    name: 'OptimisticOracleV3',
    address: '0x5953f2538F613E05bAED8A5AeFa8e6622467AD3D',
    subgraph:
      'https://subgrapher.snapshot.org/subgraph/arbitrum/7KWbDhUE5Eqcfn3LXQtLbCfJLkNucnhzJLpi2jKhqNuf',
    deployBlock: 39331673
  },
  {
    //arbitrum
    network: '42161',
    name: 'OptimisticOracleV3',
    address: '0xa6147867264374F324524E30C02C331cF28aa879',
    subgraph:
      'https://subgrapher.snapshot.org/subgraph/arbitrum/9wpkM5tHgJBHYTzKEKk4tK8a7q6MimfS9QnW7Japa8hW',
    deployBlock: 61236565
  },
  {
    // avalanche
    network: '43114',
    name: 'OptimisticOracleV3',
    address: '0xa4199d73ae206d49c966cF16c58436851f87d47F',
    subgraph:
      'https://subgrapher.snapshot.org/subgraph/arbitrum/3k8gzGzTMV2vDZAGBFM2q642SUyVbE31bAUL8SjFQkre',
    deployBlock: 27816737
  },
  {
    // core
    network: '1116',
    name: 'OptimisticOracleV3',
    address: '0xD84ACa67d683aF7702705141b3C7E57e4e5e7726',
    subgraph:
      'https://thegraph.coredao.org/subgraphs/name/umaprotocol/core-optimistic-oracle-v3',
    deployBlock: 11341063
  },
  {
    // base
    network: '8453',
    name: 'OptimisticOracleV3',
    address: '0x2aBf1Bd76655de80eDB3086114315Eec75AF500c',
    subgraph:
      'https://subgrapher.snapshot.org/subgraph/arbitrum/2Q4i8YgVZd6bAmLyDxXgrKPL2B6QwySiEUqbTyQ4vm4C',
    deployBlock: 12066343
  },
  {
    // sepolia
    network: '11155111',
    name: 'OptimisticOracleV3',
    address: '0xFd9e2642a170aDD10F53Ee14a93FcF2F31924944',
    subgraph:
      'https://subgrapher.snapshot.org/subgraph/arbitrum/78JbrMhcC9CVDZHDADvNcyhRrrccTJG4vCVBztyer1Xa',
    deployBlock: 5421195
  },
  {
    // mainnet
    network: '1',
    name: 'OptimisticGovernor',
    address: '0x28CeBFE94a03DbCA9d17143e9d2Bd1155DC26D5d',
    subgraph:
      'https://subgrapher.snapshot.org/subgraph/arbitrum/DQpwhiRSPQJEuc8y6ZBGsFfNpfwFQ8NjmjLmfv8kBkLu',
    deployBlock: 16890621
  },
  // Keep in mind, OG addresses are not the module addresses for each individual space, these addresses typically
  // are not used, but are here for reference.
  {
    //goerli
    network: '5',
    name: 'OptimisticGovernor',
    address: '0x07a7Be7AA4AaD42696A17e974486cb64A4daC47b',
    deployBlock: 8700589,
    subgraph:
      'https://api.thegraph.com/subgraphs/name/md0x/goerli-optimistic-governor'
  },
  {
    // optimism
    network: '10',
    name: 'OptimisticGovernor',
    address: '0x357fe84E438B3150d2F68AB9167bdb8f881f3b9A',
    deployBlock: 83168480,
    subgraph:
      'https://subgrapher.snapshot.org/subgraph/arbitrum/Fd5RvSfkajAJ8Mi9sPxFSMVPFf56SDivDQW3ocqTAW5'
  },
  {
    // gnosis
    network: '100',
    name: 'OptimisticGovernor',
    deployBlock: 27102135,
    subgraph:
      'https://subgrapher.snapshot.org/subgraph/arbitrum/RrkjZ6wTgLJkcjX68auzrEZHMRYwDx8kR5sFQQy4Phz'
  },
  {
    // polygon
    network: '137',
    name: 'OptimisticGovernor',
    address: '0x3Cc4b597E9c3f51288c6Cd0c087DC14c3FfdD966',
    deployBlock: 40677035,
    subgraph:
      'https://subgrapher.snapshot.org/subgraph/arbitrum/7L2JM14PnZgxGnRX7xaz54zWS6KVK6ZqVRCxEKJrJTDG'
  },
  {
    // arbitrum
    network: '42161',
    name: 'OptimisticGovernor',
    address: '0x30679ca4ea452d3df8a6c255a806e08810321763',
    deployBlock: 72850751,
    subgraph:
      'https://subgrapher.snapshot.org/subgraph/arbitrum/BfK867bnkQhnx1LspA99ypqiqxbAReQ92aZz66Ubv4tz'
  },
  {
    // avalanche
    network: '43114',
    name: 'OptimisticGovernor',
    address: '0xEF8b46765ae805537053C59f826C3aD61924Db45',
    deployBlock: 28050250,
    subgraph:
      'https://subgrapher.snapshot.org/subgraph/arbitrum/5F8875fmvtnv8Vv4aeedUcwNWjuxUg54aTHdapFuMJi3'
  },
  {
    // core
    network: '1116',
    name: 'OptimisticGovernor',
    address: '0x596Fd6A5A185c67aBD1c845b39f593fBA9C233aa',
    deployBlock: 11341122,
    subgraph:
      'https://thegraph.coredao.org/subgraphs/name/umaprotocol/core-optimistic-governor'
  },
  {
    // base
    network: '8453',
    name: 'OptimisticGovernor',
    address: '0x80bCA2E1c272239AdFDCdc87779BC8Af6E12e633',
    deployBlock: 13062540,
    subgraph:
      'https://subgrapher.snapshot.org/subgraph/arbitrum/H1WyWZqh5pHebWRDCXs7GhvGj7XznSP7arPY6pYcCqLn'
  },
  {
    // sepolia
    network: '11155111',
    name: 'OptimisticGovernor',
    address: '0x40153DdFAd90C49dbE3F5c9F96f2a5B25ec67461',
    deployBlock: 5421242,
    subgraph:
      'https://subgrapher.snapshot.org/subgraph/arbitrum/5pwrjCkpcpCd79k9MBS5yVgnsHQiw6afvXUfzqHjdRFw'
  }
] as const;

export const SAFE_APP_URLS = {
  '1': 'https://app.safe.global/apps/open',
  '5': 'https://app.safe.global/apps/open',
  '56': 'https://app.safe.global/apps/open',
  '100': 'https://app.safe.global/apps/open',
  '137': 'https://app.safe.global/apps/open',
  '246': 'https://app.safe.global/apps/open',
  '1116': 'https://safe.coredao.org/apps/open',
  '8453': 'https://app.safe.global/apps/open',
  '42161': 'https://app.safe.global/apps/open',
  '73799': 'https://app.safe.global/apps/open',
  '11155111': 'https://app.safe.global/apps/open'
} as const;

export const SAFE_PREFIXES = {
  1: 'eth',
  5: 'gor',
  10: 'oeth',
  100: 'gno',
  137: 'matic',
  1116: 'core',
  8453: 'base',
  42161: 'arb1',
  43114: 'avax',
  11155111: 'sep'
} as const;
