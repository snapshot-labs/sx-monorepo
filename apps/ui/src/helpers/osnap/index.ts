export {
  type OSnapPlugin,
  parseOSnapTransaction,
  parseInternalTransaction
} from './transactions';
export { getModuleAddressForTreasury } from './getters';
