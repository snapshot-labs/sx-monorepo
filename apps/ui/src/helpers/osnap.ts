import { ETH_CONTRACT } from './constants';
import { JsonFragment } from '@ethersproject/abi';
import {
  BaseTransaction,
  ContractCallTransaction,
  RawTransaction,
  SendNftTransaction,
  SendTokenTransaction,
  Transaction
} from '@/types';

export const transactionTypes = [
  'transferFunds',
  'transferNFT',
  'contractInteraction',
  'raw',
  'safeImport'
] as const;

type OSnapBaseTransaction = Omit<BaseTransaction, 'salt'>;

type OSnapTransferFundsTransaction = OSnapBaseTransaction & {
  type: 'transferFunds';
  amount: string;
  recipient: string;
  token: {
    name: string;
    decimals: number;
    symbol: string;
    address: string;
  };
};

type OSnapTransferNFTTransaction = OSnapBaseTransaction & {
  type: 'transferNFT';
  recipient: string;
  collectable: {
    address: string;
    id: string;
    name: string;
    tokenName: string;
  };
};

type OSnapContractInteractionTransaction = OSnapBaseTransaction & {
  type: 'contractInteraction';
  abi: string;
  method: Required<JsonFragment>;
  parameters: string[];
};

type OSnapSafeImportTransaction = OSnapBaseTransaction & {
  type: 'safeImport';
  abi: string;
  method: Required<JsonFragment>;
  parameters: { [key: string]: string };
};

type OSnapRawTransaction = OSnapBaseTransaction & {
  type: 'raw';
};

type OSnapTransaction =
  | OSnapTransferFundsTransaction
  | OSnapTransferNFTTransaction
  | OSnapContractInteractionTransaction
  | OSnapSafeImportTransaction
  | OSnapRawTransaction;

function parseTransferFundsTransaction(
  transaction: OSnapTransferFundsTransaction
): SendTokenTransaction {
  return {
    to: transaction.to,
    data: transaction.data,
    value: transaction.value,
    salt: '',
    _type: 'sendToken',
    _form: {
      recipient: transaction.recipient,
      amount: transaction.amount,
      token: {
        name: transaction.token.name,
        decimals: transaction.token.decimals,
        symbol: transaction.token.symbol,
        address: transaction.token.address === 'main' ? ETH_CONTRACT : transaction.token.address
      }
    }
  };
}

function parseTransferNFTTransaction(transaction: OSnapTransferNFTTransaction): SendNftTransaction {
  return {
    to: transaction.to,
    data: transaction.data,
    value: transaction.value,
    salt: '',
    _type: 'sendNft',
    _form: {
      recipient: transaction.recipient,
      amount: '1',
      nft: {
        address: transaction.collectable.address,
        id: transaction.collectable.id,
        name: transaction.collectable.name,
        collection: transaction.collectable.tokenName
      }
    }
  };
}

function parseContractInteractionTransaction(
  transaction: OSnapContractInteractionTransaction
): ContractCallTransaction {
  return {
    to: transaction.to,
    data: transaction.data,
    value: transaction.value,
    salt: '',
    _type: 'contractCall',
    _form: {
      recipient: transaction.to,
      method: transaction.method.name,
      args: transaction.parameters,
      abi: JSON.parse(transaction.abi)
    }
  };
}

function parseSafeImportTransaction(
  transaction: OSnapSafeImportTransaction
): ContractCallTransaction {
  return {
    to: transaction.to,
    data: transaction.data,
    value: transaction.value,
    salt: '',
    _type: 'contractCall',
    _form: {
      recipient: transaction.to,
      method: transaction.method.name,
      args: transaction.method.inputs.map(input =>
        input.name ? transaction.parameters[input.name] : null
      ),
      abi: JSON.parse(transaction.abi)
    }
  };
}

function parseRawTransaction(transaction: OSnapRawTransaction): RawTransaction {
  return {
    to: transaction.to,
    data: transaction.data,
    value: transaction.value,
    salt: '',
    _type: 'raw',
    _form: {
      recipient: transaction.to
    }
  };
}

export function parseOSnapTransaction(transaction: OSnapTransaction): Transaction {
  switch (transaction.type) {
    case 'transferFunds':
      return parseTransferFundsTransaction(transaction);
    case 'transferNFT':
      return parseTransferNFTTransaction(transaction);
    case 'contractInteraction':
      return parseContractInteractionTransaction(transaction);
    case 'safeImport':
      return parseSafeImportTransaction(transaction);
    case 'raw':
      return parseRawTransaction(transaction);
    default:
      throw new Error('Invalid transaction type');
  }
}
