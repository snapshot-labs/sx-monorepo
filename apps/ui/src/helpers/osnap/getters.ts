import { ApolloClient, gql, InMemoryCache } from '@apollo/client/core';
import { defaultAbiCoder } from '@ethersproject/abi';
import { keccak256 } from '@ethersproject/keccak256';
import { pack } from '@ethersproject/solidity';
import { toUtf8Bytes } from '@ethersproject/strings';
import { Transaction } from '@/types';
import { contractData, SAFE_APP_URLS, SAFE_PREFIXES } from './constants';

type SpaceConfigResponse =
  | {
      automaticExecution: true;
    }
  | {
      automaticExecution: false;
      rules: boolean;
      bondToken: boolean;
      bondAmount: boolean;
    };

/**
 * The `proposalHash` as represented in the Optimistic Governor contract is the keccak256 hash of the transactions that make up the proposal.
 */
export function getProposalHashFromTransactions(transactions: Transaction[]) {
  const rawTxs = transactions.map(tx => {
    return {
      to: tx.to,
      operation: 0,
      value: tx.value,
      data: tx.data
    };
  });

  return keccak256(
    defaultAbiCoder.encode(
      ['(address to, uint8 operation, uint256 value, bytes data)[]'],
      [rawTxs]
    )
  );
}

/**
 * Fetches the subgraph url for a given contract on a given network.
 */
export function getContractSubgraph(params: { chainId: number; name: string }) {
  const result = contractData.find(
    contract =>
      contract.network === params.chainId.toString() &&
      contract.name === params.name
  );

  if (!result)
    throw new Error(
      `No subgraph url defined for ${params.name} on network ${params.chainId}`
    );

  return result.subgraph;
}

/**
 * A helper that wraps the getContractSubgraph function to return the subgraph url for the OptimisticGovernor contract on a given network.
 */
export function getOptimisticGovernorSubgraph(chainId: number): string {
  return getContractSubgraph({ chainId, name: 'OptimisticGovernor' });
}

/**
 * Returns the address of the Optimistic Governor contract deployment associated with a given treasury (Safe) from the graph.
 */
export const getModuleAddressForTreasury = async (
  chainId: number,
  treasuryAddress: string
) => {
  try {
    const subgraph = getOptimisticGovernorSubgraph(chainId);
    const query = gql`
      query ($treasuryAddress: String!) {
        safe(id: $treasuryAddress) {
          optimisticGovernor {
            id
          }
        }
      }
    `;

    type Result = {
      safe?: { optimisticGovernor?: { id: string } };
    };

    const client = new ApolloClient({
      uri: subgraph,
      cache: new InMemoryCache()
    });

    const { data } = await client.query({
      query,
      variables: {
        treasuryAddress: treasuryAddress.toLowerCase()
      }
    });

    return (data as Result).safe?.optimisticGovernor?.id ?? '';
  } catch (error) {
    console.warn(
      `Unable to get module address for treasury ${treasuryAddress} on network ${chainId}`
    );

    throw error;
  }
};

/**
 * Checks if a given treasury (Safe) has enabled oSnap.
 */
export const getIsOsnapEnabled = async (
  chainId: number,
  treasuryAddress: string
) => {
  try {
    const subgraph = getOptimisticGovernorSubgraph(chainId);
    const query = gql`
      query ($treasuryAddress: String!) {
        safe(id: $treasuryAddress) {
          isOptimisticGovernorEnabled
        }
      }
    `;

    type Result = {
      safe?: { isOptimisticGovernorEnabled?: boolean };
    };

    const client = new ApolloClient({
      uri: subgraph,
      cache: new InMemoryCache()
    });

    const { data } = await client.query({
      query,
      variables: {
        treasuryAddress: treasuryAddress.toLowerCase()
      }
    });

    return (data as Result).safe?.isOptimisticGovernorEnabled ?? false;
  } catch (error) {
    console.warn(
      `Unable to check if oSnap is enable for address ${treasuryAddress} on network ${chainId}`
    );

    throw error;
  }
};

/**
 * Fetches the details of a given proposal from the Optimistic Governor subgraph.
 *
 * The subgraph uses the `assertionId` that comes from assertion events as the primary key for proposals.
 * However, this `assertionId` will be deleted if the proposal is disputed, so we can't use it to query the subgraph.
 * Instead, we use the `proposalHash` and `explanation` to query the subgraph.
 * The `explanation` contains the ipfs url of the proposal, which is the only way to distinguish between proposals with the same `proposalHash`.
 * This means we must use a `where` clause to filter the results, which is not ideal.
 */
export async function getOgProposalGql(params: {
  chainId: number;
  explanation: string;
  moduleAddress: string;
  proposalHash: string;
}) {
  const { chainId, explanation, moduleAddress, proposalHash } = params;
  const encodedExplanation = pack(
    ['bytes'],
    [toUtf8Bytes(explanation.replace(/^0x/, ''))]
  );

  const subgraph = getOptimisticGovernorSubgraph(chainId);
  const query = gql`
    query (
      $proposalHash: String!
      $explanation: String!
      $moduleAddress: String!
    ) {
      proposals(
        where: {
          proposalHash: $proposalHash
          explanation: $explanation
          optimisticGovernor: $moduleAddress
        }
      ) {
        id
        executed
        deleted
        assertionId
        executionTransactionHash
      }
    }
  `;

  type Result = {
    proposals: {
      id: string;
      executed: boolean;
      assertionId: string;
      deleted: boolean;
      executionTransactionHash: string;
    }[];
  };

  const client = new ApolloClient({
    uri: subgraph,
    cache: new InMemoryCache()
  });

  const { data } = await client.query({
    query,
    variables: {
      proposalHash,
      explanation: encodedExplanation,
      moduleAddress
    }
  });

  return (data as Result).proposals[0];
}

/**
 * Check if a space's deployed (on-chain) settings are supported by oSnap bots for auto execution
 */
export async function isConfigCompliant(safeAddress: string, chainId: number) {
  const res = await fetch(
    `https://osnap.uma.xyz/api/space-config?address=${safeAddress}&chainId=${chainId}`
  );

  if (!res.ok) {
    throw new Error('Unable to fetch setting status');
  }
  const data = await res.json();
  return data as unknown as SpaceConfigResponse;
}

/**
 * Returns the EIP-3770 prefix for a given network.
 *
 * @see SafeNetworkPrefix
 */
export function getSafeNetworkPrefix(chainId: number) {
  return SAFE_PREFIXES[chainId];
}

/**
 * Creates the url for the Safe app to configure oSnap.
 *
 * The data that the Safe app needs is encoded as URL search params.
 */
export function makeConfigureOsnapUrl(params: {
  safeAddress: string;
  network: number;
  spaceName: string;
  spaceUrl: string;
  baseUrl?: string;
  appUrl?: string;
}) {
  const {
    safeAddress,
    network,
    spaceName,
    spaceUrl,
    appUrl = 'https://osnap.uma.xyz/'
  } = params;
  const baseUrl =
    params.baseUrl ??
    SAFE_APP_URLS[network] ??
    'https://app.safe.global/apps/open';
  const safeAddressPrefix = getSafeNetworkPrefix(network);
  const appUrlSearchParams = new URLSearchParams();
  appUrlSearchParams.set('spaceName', spaceName);
  appUrlSearchParams.set('spaceUrl', spaceUrl);
  const appUrlSearch = appUrlSearchParams.toString();
  const safeAppSearchParams = new URLSearchParams();
  safeAppSearchParams.set('safe', `${safeAddressPrefix}:${safeAddress}`);
  safeAppSearchParams.set('appUrl', `${appUrl}?${appUrlSearch}`);
  const safeAppSearch = safeAppSearchParams.toString();
  const url = `${baseUrl}?${safeAppSearch}`;
  return url;
}
