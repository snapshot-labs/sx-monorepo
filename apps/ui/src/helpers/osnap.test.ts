import { describe, it, expect } from 'vitest';
import { parseOSnapTransaction } from './osnap';

describe('parseOSnapTransaction', () => {
  it('should parse oSnap transfer funds transaction', () => {
    const transaction = {
      to: '0x556B14CbdA79A36dC33FcD461a04A5BCb5dC2A70',
      data: '0x',
      type: 'transferFunds' as const,
      token: {
        name: 'Ether',
        symbol: 'ETH',
        address: 'main',
        balance: '0.379',
        chainId: '11155111',
        logoUri: 'https://safe-transaction-assets.safe.global/chains/1/currency_logo.png',
        decimals: 18,
        verified: false
      },
      value: '1000000000000000',
      amount: '1000000000000000',
      isValid: true,
      formatted: ['0x556B14CbdA79A36dC33FcD461a04A5BCb5dC2A70', 0, '1000000000000000', '0x'],
      recipient: '0x556B14CbdA79A36dC33FcD461a04A5BCb5dC2A70'
    };

    expect(parseOSnapTransaction(transaction)).toMatchSnapshot();
  });

  it('should parse oSnap transfer NFT transaction', () => {
    const transaction = {
      to: '0x5A96CF3ace257Dfcc1fd3C037e548585124dc0C5',
      data: '0x42842e0e0000000000000000000000008edfcc5f141ffc2b6892530d1fb21bbcdc74b455000000000000000000000000556b14cbda79a36dc33fcd461a04a5bcb5dc2a70000000000000000000000000000000000000000000000000000000000000032a',
      type: 'transferNFT' as const,
      value: '0',
      isValid: true,
      formatted: [
        '0x5A96CF3ace257Dfcc1fd3C037e548585124dc0C5',
        0,
        '0',
        '0x42842e0e0000000000000000000000008edfcc5f141ffc2b6892530d1fb21bbcdc74b455000000000000000000000000556b14cbda79a36dc33fcd461a04a5bcb5dc2a70000000000000000000000000000000000000000000000000000000000000032a'
      ],
      recipient: '0x556B14CbdA79A36dC33FcD461a04A5BCb5dC2A70',
      collectable: {
        id: '810',
        uri: 'https://bafkreigtcra4mp2u5ppor2akc5mxjop6ag4o72ryugic63goqbxscpwygy.ipfs.nftstorage.link/',
        name: 'Weeedidit Palls #101',
        address: '0x5A96CF3ace257Dfcc1fd3C037e548585124dc0C5',
        logoUri:
          'https://safe-transaction-assets.safe.global/tokens/logos/0x5A96CF3ace257Dfcc1fd3C037e548585124dc0C5.png',
        imageUri:
          'https://nft-delivery.infura-ipfs.io/ipfs/QmQiJyvrV8sETi8HtV89yaJit5or7ZaMUo8pFsCJXMM8RL/101.png',
        metadata: {
          name: 'Weeedidit Palls #101',
          image:
            'https://nft-delivery.infura-ipfs.io/ipfs/QmQiJyvrV8sETi8HtV89yaJit5or7ZaMUo8pFsCJXMM8RL/101.png',
          attributes: [],
          description:
            'Step into a surreal universe of vibrant colors and mind-bending shapes. Explore a world where imagination reigns and creativity knows no bounds.'
        },
        tokenName: 'Weee Did It Palz',
        description:
          'Step into a surreal universe of vibrant colors and mind-bending shapes. Explore a world where imagination reigns and creativity knows no bounds.',
        tokenSymbol: 'WDIT'
      }
    };

    expect(parseOSnapTransaction(transaction)).toMatchSnapshot();
  });

  it('should parse oSnap contract interaction transaction', () => {
    const transaction = {
      to: '0x000000000000cd17345801aa8147b8D3950260FF',
      abi: '[{"anonymous":false,"inputs":[{"indexed":true,"internalType":"address","name":"user","type":"address"},{"indexed":false,"internalType":"string","name":"content","type":"string"},{"indexed":true,"internalType":"string","name":"tag","type":"string"}],"name":"NewPost","type":"event"},{"inputs":[{"internalType":"string","name":"content","type":"string"},{"internalType":"string","name":"tag","type":"string"}],"name":"post","outputs":[],"stateMutability":"nonpayable","type":"function"}]',
      data: '0x0ae1b13d000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000800000000000000000000000000000000000000000000000000000000000000004746573740000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000047465737400000000000000000000000000000000000000000000000000000000',
      type: 'contractInteraction' as const,
      value: '0',
      method: {
        gas: null,
        name: 'post',
        type: 'function',
        inputs: [
          {
            name: 'content',
            type: 'string',
            indexed: null,
            baseType: 'string',
            components: null,
            arrayLength: null,
            _isParamType: true,
            arrayChildren: null
          },
          {
            name: 'tag',
            type: 'string',
            indexed: null,
            baseType: 'string',
            components: null,
            arrayLength: null,
            _isParamType: true,
            arrayChildren: null
          }
        ],
        outputs: [],
        payable: false,
        constant: false,
        _isFragment: true,
        stateMutability: 'nonpayable'
      } as any,
      isValid: true,
      formatted: [
        '0x000000000000cd17345801aa8147b8D3950260FF',
        0,
        '0',
        '0x0ae1b13d000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000800000000000000000000000000000000000000000000000000000000000000004746573740000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000047465737400000000000000000000000000000000000000000000000000000000'
      ],
      parameters: ['test', 'test']
    };

    expect(parseOSnapTransaction(transaction)).toMatchSnapshot();
  });

  it('should parse oSnap safeImport transaction', () => {
    const transaction = {
      to: '0x4F604735c1cF31399C6E711D5962b2B3E0225AD3',
      abi: '[{"type":"function","name":"transfer","constant":false,"inputs":[{"name":"to","type":"address","indexed":null,"components":null,"arrayLength":null,"arrayChildren":null,"baseType":"address","_isParamType":true},{"name":"amount","type":"uint256","indexed":null,"components":null,"arrayLength":null,"arrayChildren":null,"baseType":"uint256","_isParamType":true}],"outputs":[],"payable":false,"stateMutability":"nonpayable","gas":null,"_isFragment":true}]',
      data: '0xa9059cbb000000000000000000000000fe1552da65facaac5b50b73ceda4c993e16d46940000000000000000000000000000000000000000000000000de0b6b3a7640000',
      type: 'safeImport' as const,
      value: '0',
      method: {
        gas: null,
        name: 'transfer',
        type: 'function',
        inputs: [
          {
            name: 'to',
            type: 'address',
            indexed: null,
            baseType: 'address',
            components: null,
            arrayLength: null,
            _isParamType: true,
            arrayChildren: null
          },
          {
            name: 'amount',
            type: 'uint256',
            indexed: null,
            baseType: 'uint256',
            components: null,
            arrayLength: null,
            _isParamType: true,
            arrayChildren: null
          }
        ],
        outputs: [],
        payable: false,
        constant: false,
        _isFragment: true,
        stateMutability: 'nonpayable'
      } as any,
      isValid: true,
      formatted: [
        '0x4F604735c1cF31399C6E711D5962b2B3E0225AD3',
        0,
        '0',
        '0xa9059cbb000000000000000000000000fe1552da65facaac5b50b73ceda4c993e16d46940000000000000000000000000000000000000000000000000de0b6b3a7640000'
      ],
      parameters: {
        to: '0xfE1552DA65FAcAaC5B50b73CEDa4C993e16d4694',
        amount: '1000000000000000000'
      }
    };

    expect(parseOSnapTransaction(transaction)).toMatchSnapshot();
  });

  it('should parse oSnap raw transaction', () => {
    const transaction = {
      to: '0x556B14CbdA79A36dC33FcD461a04A5BCb5dC2A70',
      data: '0x',
      type: 'raw' as const,
      value: '1',
      isValid: true,
      formatted: ['0x556B14CbdA79A36dC33FcD461a04A5BCb5dC2A70', 0, '1', '0x']
    };

    expect(parseOSnapTransaction(transaction)).toMatchSnapshot();
  });

  it('should throw on unknown transaction', () => {
    const transaction = {
      type: 'unknown'
    };

    expect(() => parseOSnapTransaction(transaction as any)).toThrowError(
      'Invalid transaction type'
    );
  });
});
