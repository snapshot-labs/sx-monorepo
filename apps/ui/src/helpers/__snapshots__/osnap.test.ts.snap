// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parseOSnapTransaction > should parse oSnap contract interaction transaction 1`] = `
{
  "_form": {
    "abi": [
      {
        "anonymous": false,
        "inputs": [
          {
            "indexed": true,
            "internalType": "address",
            "name": "user",
            "type": "address",
          },
          {
            "indexed": false,
            "internalType": "string",
            "name": "content",
            "type": "string",
          },
          {
            "indexed": true,
            "internalType": "string",
            "name": "tag",
            "type": "string",
          },
        ],
        "name": "NewPost",
        "type": "event",
      },
      {
        "inputs": [
          {
            "internalType": "string",
            "name": "content",
            "type": "string",
          },
          {
            "internalType": "string",
            "name": "tag",
            "type": "string",
          },
        ],
        "name": "post",
        "outputs": [],
        "stateMutability": "nonpayable",
        "type": "function",
      },
    ],
    "args": [
      "test",
      "test",
    ],
    "method": "post",
    "recipient": "0x000000000000cd17345801aa8147b8D3950260FF",
  },
  "_type": "contractCall",
  "data": "0x0ae1b13d000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000800000000000000000000000000000000000000000000000000000000000000004746573740000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000047465737400000000000000000000000000000000000000000000000000000000",
  "salt": "",
  "to": "0x000000000000cd17345801aa8147b8D3950260FF",
  "value": "0",
}
`;

exports[`parseOSnapTransaction > should parse oSnap raw transaction 1`] = `
{
  "_form": {
    "recipient": "0x556B14CbdA79A36dC33FcD461a04A5BCb5dC2A70",
  },
  "_type": "raw",
  "data": "0x",
  "salt": "",
  "to": "0x556B14CbdA79A36dC33FcD461a04A5BCb5dC2A70",
  "value": "1",
}
`;

exports[`parseOSnapTransaction > should parse oSnap safeImport transaction 1`] = `
{
  "_form": {
    "abi": [
      {
        "_isFragment": true,
        "constant": false,
        "gas": null,
        "inputs": [
          {
            "_isParamType": true,
            "arrayChildren": null,
            "arrayLength": null,
            "baseType": "address",
            "components": null,
            "indexed": null,
            "name": "to",
            "type": "address",
          },
          {
            "_isParamType": true,
            "arrayChildren": null,
            "arrayLength": null,
            "baseType": "uint256",
            "components": null,
            "indexed": null,
            "name": "amount",
            "type": "uint256",
          },
        ],
        "name": "transfer",
        "outputs": [],
        "payable": false,
        "stateMutability": "nonpayable",
        "type": "function",
      },
    ],
    "args": [
      "0xfE1552DA65FAcAaC5B50b73CEDa4C993e16d4694",
      "1000000000000000000",
    ],
    "method": "transfer",
    "recipient": "0x4F604735c1cF31399C6E711D5962b2B3E0225AD3",
  },
  "_type": "contractCall",
  "data": "0xa9059cbb000000000000000000000000fe1552da65facaac5b50b73ceda4c993e16d46940000000000000000000000000000000000000000000000000de0b6b3a7640000",
  "salt": "",
  "to": "0x4F604735c1cF31399C6E711D5962b2B3E0225AD3",
  "value": "0",
}
`;

exports[`parseOSnapTransaction > should parse oSnap transfer NFT transaction 1`] = `
{
  "_form": {
    "amount": "1",
    "nft": {
      "address": "0x5A96CF3ace257Dfcc1fd3C037e548585124dc0C5",
      "collection": "Weee Did It Palz",
      "id": "810",
      "name": "Weeedidit Palls #101",
    },
    "recipient": "0x556B14CbdA79A36dC33FcD461a04A5BCb5dC2A70",
  },
  "_type": "sendNft",
  "data": "0x42842e0e0000000000000000000000008edfcc5f141ffc2b6892530d1fb21bbcdc74b455000000000000000000000000556b14cbda79a36dc33fcd461a04a5bcb5dc2a70000000000000000000000000000000000000000000000000000000000000032a",
  "salt": "",
  "to": "0x5A96CF3ace257Dfcc1fd3C037e548585124dc0C5",
  "value": "0",
}
`;

exports[`parseOSnapTransaction > should parse oSnap transfer funds transaction 1`] = `
{
  "_form": {
    "amount": "1000000000000000",
    "recipient": "0x556B14CbdA79A36dC33FcD461a04A5BCb5dC2A70",
    "token": {
      "address": "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee",
      "decimals": 18,
      "name": "Ether",
      "symbol": "ETH",
    },
  },
  "_type": "sendToken",
  "data": "0x",
  "salt": "",
  "to": "0x556B14CbdA79A36dC33FcD461a04A5BCb5dC2A70",
  "value": "1000000000000000",
}
`;
